src/OrchardCore/OrchardCore.Mvc.Core/LocationExpander/ModularViewLocationExpanderProvider.cs

```diff
@@ -54,7 +71,62 @@ IEnumerable<string> PageViewLocations()
 
             var extensionViewsPath = '/' + extension.SubPath + "/Views";
             result.Add(extensionViewsPath + "/{1}/{0}" + RazorViewEngine.ViewExtension);
-            result.Add(extensionViewsPath + "/Shared/{0}" + RazorViewEngine.ViewExtension);
+
+            if (!context.ViewName.StartsWith("Components/", StringComparison.Ordinal))
+            {
+                result.Add(extensionViewsPath + "/Shared/{0}" + RazorViewEngine.ViewExtension);
+            }
+            else
+            {
+                if (!_memoryCache.TryGetValue(CacheKey, out IDictionary<string,
+                    IEnumerable<string>> moduleViewLocationsByViewComponentName))
+                {
+                    var modules = _extensionManager.GetFeatures()
+                        .Where(f => f.Id == f.Extension.Id &&
+                            _shellDescriptor.Features.Any(sf => sf.Id == f.Id))
+                        .Select(f => f.Extension);
+
+                    var feature = new ViewComponentFeature();
+                    _applicationPartManager.PopulateFeature(feature);
+
+                    var modulesByComponentName = new Dictionary<string, IList<IExtensionInfo>>();
+
+                    foreach (var component in feature.ViewComponents.Select(vc => vc.AsType()))
+                    {
+                        var module = modules.FirstOrDefault(e => e.Id == component.Assembly.GetName().Name);
+
+                        if (module == null)
+                        {
+                            continue;
+                        }
+
+                        if (!modulesByComponentName.ContainsKey(component.Name))
+                        {
+                            modulesByComponentName[component.Name] = new List<IExtensionInfo>();
+                        }
+
+                        modulesByComponentName[component.Name].Add(module);
+                    }
+
+                    var orderedModuleNames = modules.Select(e => e.Id).Reverse().ToList();
+
+                    moduleViewLocationsByViewComponentName = modulesByComponentName
+                        .ToDictionary(kvp => kvp.Key, kvp => kvp.Value
+                            .OrderBy(module => orderedModuleNames.IndexOf(module.Id))
+                            .Select(module => '/' + module.SubPath + "/Views" + "/Shared/{0}"
+                                + RazorViewEngine.ViewExtension));
+
+                    _memoryCache.Set(CacheKey, moduleViewLocationsByViewComponentName);
+                }
+
+                var viewComponentName = context.ViewName.Substring(0, context.ViewName.LastIndexOf('/'))
```

Comment:
Can this code be more complex?

Reply:
Yes it could be ;) by not only relying on the same VC name but also check for types compatibility.
No, i will re-take a look on this to try to make it simpler, if you have any idea. Meanwhile, here:

- So i get only the main features of enabled & ordered modules.

- Then, to retrieve existing VC from the part manager, it is as mvc do.

- Then, then ... I create and cache a dictionary whose key is a VC name and value is a list of view locations related to the modules which define a VC with this name. And the list needs to be ordered according to modules deps.

Before i was using a simpler solution but which was adding many locations (one per module), here it's more complex but most of the time we only add one location. But as said above, i will re-take a look on this.